File: mobile-app-sdk/android/JarvisSDK.kt
package com.jarvis.sdk

import android.content.Context
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.core.content.PermissionChecker
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

// MARK: - Models

data class JarvisConfig(
    val baseURL: String,
    val apiKey: String,
    val timeout: Long = 30L,
    val retryAttempts: Int = 3,
    val enableLogging: Boolean = false
)

data class DeviceInfo(
    @SerializedName("device_id") val deviceID: String,
    val platform: String,
    @SerializedName("app_version") val appVersion: String,
    @SerializedName("os_version") val osVersion: String,
    @SerializedName("device_model") val deviceModel: String,
    val timezone: String
)

data class AuthRequest(
    @SerializedName("device_info") val deviceInfo: DeviceInfo
)

data class AuthResponse(
    val success: Boolean,
    @SerializedName("user_id") val userID: String,
    @SerializedName("mobile_token") val token: String,
    @SerializedName("session_id") val sessionID: String,
    @SerializedName("expires_at") val expiresAt: String,
    @SerializedName("user_info") val userInfo: Map<String, Any>
)

data class VoiceRequest(
    @SerializedName("session_id") val sessionID: String,
    @SerializedName("audio_data") val audioData: String? = null,
    val text: String? = null,
    @SerializedName("voice_settings") val voiceSettings: VoiceSettings? = null,
    val language: String = "en",
    val format: String = "wav"
)

data class VoiceSettings(
    val voice: String = "en-US-AriaNeural"
)

data class VoiceResponse(
    val success: Boolean,
    @SerializedName("response_text") val text: String?,
    @SerializedName("response_audio") val audioData: String?,
    @SerializedName("processing_time") val processingTime: Double,
    val confidence: Double,
    val language: String
)

data class CommandRequest(
    @SerializedName("session_id") val sessionID: String,
    val command: String,
    val language: String = "en",
    @SerializedName("send_notification") val sendNotification: Boolean = false
)

data class CommandResponse(
    val success: Boolean,
    val command: String,
    val response: String,
    val confidence: Double,
    @SerializedName("processing_time") val processingTime: Double
)

// MARK: - Exceptions

sealed class JarvisException(message: String) : Exception(message) {
    class AuthenticationFailed(message: String) : JarvisException("Authentication failed: $message")
    class NetworkError(message: String) : JarvisException("Network error: $message")
    class VoiceProcessingError(message: String) : JarvisException("Voice processing error: $message")
    class ConfigurationError(message: String) : JarvisException("Configuration error: $message")
    class PermissionDenied(message: String) : JarvisException("Permission denied: $message")
    class SessionExpired : JarvisException("Session expired")
    class InvalidResponse : JarvisException("Invalid response from server")
}

// MARK: - Result Wrapper

sealed class Result<T> {
    data class Success<T>(val data: T) : Result<T>()
    data class Error<T>(val error: JarvisException) : Result<T>()
}

// MARK: - Main SDK Class

class JarvisClient(private val context: Context) {
    private var config: JarvisConfig? = null
    private var authToken: String? = null
    private var sessionID: String? = null
    private var userID: String? = null
    
    private val httpClient: OkHttpClient
    private val gson = Gson()
    private val mainHandler = Handler(Looper.getMainLooper())
    
    // Audio recording
    private var audioRecord: AudioRecord? = null
    private var isRecording = false
    private var recordingThread: Thread? = null
    private var voiceSessionActive = false
    
    // Audio configuration
    private val sampleRate = 16000
    private val channelConfig = AudioFormat.CHANNEL_IN_MONO
    private val audioFormat = AudioFormat.ENCODING_PCM_16BIT
    private val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat)
    
    companion object {
        private const val TAG = "JarvisSDK"
        private const val PERMISSION_RECORD_AUDIO = android.Manifest.permission.RECORD_AUDIO
    }
    
    init {
        httpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()
    }
    
    // MARK: - Configuration
    
    fun configure(config: JarvisConfig) {
        this.config = config
        if (config.enableLogging) {
            Log.d(TAG, "Jarvis SDK configured with base URL: ${config.baseURL}")
        }
    }
    
    // MARK: - Authentication
    
    suspend fun authenticate(deviceInfo: DeviceInfo): Result<AuthResponse> {
        val config = this.config ?: return Result.Error(JarvisException.ConfigurationError("SDK not configured"))
        
        val url = "${config.baseURL}/api/v2/mobile/auth/login"
        val request = AuthRequest(deviceInfo)
        
        return try {
            val response = makeRequest<AuthResponse>(url, request, config.apiKey)
            
            response.data?.let { authResponse ->
                if (authResponse.success) {
                    // Store authentication data
                    authToken = authResponse.token
                    sessionID = authResponse.sessionID
                    userID = authResponse.userID
                    
                    Result.Success(authResponse)
                } else {
                    Result.Error(JarvisException.AuthenticationFailed("Authentication failed"))
                }
            } ?: Result.Error(JarvisException.InvalidResponse())
            
        } catch (e: Exception) {
            Result.Error(JarvisException.NetworkError(e.message ?: "Unknown error"))
        }
    }
    
    // MARK: - Voice Processing
    
    suspend fun startVoiceSession(): Result<VoiceResponse> {
        if (authToken == null) {
            return Result.Error(JarvisException.AuthenticationFailed("Not authenticated"))
        }
        
        // Check microphone permission
        if (!hasMicrophonePermission()) {
            return Result.Error(JarvisException.PermissionDenied("Microphone access denied"))
        }
        
        return try {
            startRecording()
            Result.Success(VoiceResponse(
                success = true,
                text = "Voice session started",
                audioData = null,
                processingTime = 0.0,
                confidence = 1.0,
                language = "en"
            ))
        } catch (e: Exception) {
            Result.Error(JarvisException.VoiceProcessingError(e.message ?: "Failed to start voice session"))
        }
    }
    
    private fun hasMicrophonePermission(): Boolean {
        return ContextCompat.checkSelfPermission(context, PERMISSION_RECORD_AUDIO) == PermissionChecker.PERMISSION_GRANTED
    }
    
    private fun startRecording() {
        if (isRecording) return
        
        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            channelConfig,
            audioFormat,
            bufferSize
        )
        
        audioRecord?.startRecording()
        isRecording = true
        voiceSessionActive = true
        
        recordingThread = Thread {
            val buffer = ByteArray(bufferSize)
            
            while (isRecording) {
                val bytesRead = audioRecord?.read(buffer, 0, bufferSize) ?: 0
                if (bytesRead > 0) {
                    // Process audio buffer
                    processAudioBuffer(buffer.copyOf(bytesRead))
                }
            }
        }
        
        recordingThread?.start()
        
        if (config?.enableLogging == true) {
            Log.d(TAG, "Voice recording started")
        }
    }
    
    private fun processAudioBuffer(buffer: ByteArray) {
        // Convert audio buffer to base64
        val base64Audio = Base64.encodeToString(buffer, Base64.DEFAULT)
        
        // Send to server for processing
        GlobalScope.launch {
            processVoiceData(base64Audio)
        }
    }
    
    private suspend fun processVoiceData(audioData: String): Result<VoiceResponse> {
        val config = this.config ?: return Result.Error(JarvisException.ConfigurationError("SDK not configured"))
        val sessionID = this.sessionID ?: return Result.Error(JarvisException.SessionExpired())
        
        val url = "${config.baseURL}/api/v2/mobile/voice/process"
        val request = VoiceRequest(
            sessionID = sessionID,
            audioData = audioData,
            language = "en",
            format = "wav"
        )
        
        return try {
            val response = makeAuthenticatedRequest<VoiceResponse>(url, request)
            response
        } catch (e: Exception) {
            Result.Error(JarvisException.VoiceProcessingError(e.message ?: "Voice processing failed"))
        }
    }
    
    fun stopVoiceSession() {
        if (isRecording) {
            isRecording = false
            voiceSessionActive = false
            
            audioRecord?.stop()
            audioRecord?.release()
            audioRecord = null
            
            recordingThread?.join()
            recordingThread = null
            
            if (config?.enableLogging == true) {
                Log.d(TAG, "Voice recording stopped")
            }
        }
    }
    
    // MARK: - Command Processing
    
    suspend fun executeCommand(command: String): Result<CommandResponse> {
        val config = this.config ?: return Result.Error(JarvisException.ConfigurationError("SDK not configured"))
        val sessionID = this.sessionID ?: return Result.Error(JarvisException.SessionExpired())
        
        val url = "${config.baseURL}/api/v2/mobile/command"
        val request = CommandRequest(
            sessionID = sessionID,
            command = command,
            language = "en",
            sendNotification = false
        )
        
        return try {
            val response = makeAuthenticatedRequest<CommandResponse>(url, request)
            response
        } catch (e: Exception) {
            Result.Error(JarvisException.NetworkError(e.message ?: "Command execution failed"))
        }
    }
    
    // MARK: - Text-to-Speech
    
    suspend fun synthesizeSpeech(text: String, voice: String = "en-US-AriaNeural"): Result<VoiceResponse> {
        val config = this.config ?: return Result.Error(JarvisException.ConfigurationError("SDK not configured"))
        val sessionID = this.sessionID ?: return Result.Error(JarvisException.SessionExpired())
        
        val url = "${config.baseURL}/api/v2/mobile/voice/process"
        val request = VoiceRequest(
            sessionID = sessionID,
            text = text,
            voiceSettings = VoiceSettings(voice = voice),
            language = "en"
        )
        
        return try {
            val response = makeAuthenticatedRequest<VoiceResponse>(url, request)
            response
        } catch (e: Exception) {
            Result.Error(JarvisException.VoiceProcessingError(e.message ?: "TTS processing failed"))
        }
    }
    
    // MARK: - HTTP Utilities
    
    private suspend inline fun <reified T> makeRequest(url: String, requestBody: Any, apiKey: String): Result<T> {
        return withContext(Dispatchers.IO) {
            try {
                val json = gson.toJson(requestBody)
                val body = json.toRequestBody("application/json".toMediaType())
                
                val request = Request.Builder()
                    .url(url)
                    .post(body)
                    .addHeader("Content-Type", "application/json")
                    .addHeader("Authorization", "Bearer $apiKey")
                    .build()
                
                val response = httpClient.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val responseBody = response.body?.string()
                    if (responseBody != null) {
                        val result = gson.fromJson(responseBody, T::class.java)
                        Result.Success(result)
                    } else {
                        Result.Error(JarvisException.InvalidResponse())
                    }
                } else {
                    Result.Error(JarvisException.NetworkError("HTTP ${response.code}"))
                }
            } catch (e: IOException) {
                Result.Error(JarvisException.NetworkError(e.message ?: "Network request failed"))
            } catch (e: Exception) {
                Result.Error(JarvisException.NetworkError(e.message ?: "Unknown error"))
            }
        }
    }
    
    private suspend inline fun <reified T> makeAuthenticatedRequest(url: String, requestBody: Any): Result<T> {
        val authToken = this.authToken ?: return Result.Error(JarvisException.AuthenticationFailed("Not authenticated"))
        
        return withContext(Dispatchers.IO) {
            try {
                val json = gson.toJson(requestBody)
                val body = json.toRequestBody("application/json".toMediaType())
                
                val request = Request.Builder()
                    .url(url)
                    .post(body)
                    .addHeader("Content-Type", "application/json")
                    .addHeader("Authorization", "Bearer $authToken")
                    .build()
                
                val response = httpClient.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val responseBody = response.body?.string()
                    if (responseBody != null) {
                        val result = gson.fromJson(responseBody, T::class.java)
                        Result.Success(result)
                    } else {
                        Result.Error(JarvisException.InvalidResponse())
                    }
                } else {
                    Result.Error(JarvisException.NetworkError("HTTP ${response.code}"))
                }
            } catch (e: IOException) {
                Result.Error(JarvisException.NetworkError(e.message ?: "Network request failed"))
            } catch (e: Exception) {
                Result.Error(JarvisException.NetworkError(e.message ?: "Unknown error"))
            }
        }
    }
    
    // MARK: - Utility Methods
    
    fun isAuthenticated(): Boolean {
        return authToken != null && userID != null
    }
    
    fun getCurrentUserID(): String? {
        return userID
    }
    
    fun getCurrentSessionID(): String? {
        return sessionID
    }
    
    fun logout() {
        authToken = null
        sessionID = null
        userID = null
        stopVoiceSession()
    }
    
    // MARK: - Cleanup
    
    fun cleanup() {
        stopVoiceSession()
        httpClient.dispatcher.executorService.shutdown()
    }
}

// MARK: - Extensions

fun JarvisClient.executeCommandAsync(command: String, callback: (Result<CommandResponse>) -> Unit) {
    GlobalScope.launch {
        val result = executeCommand(command)
        withContext(Dispatchers.Main) {
            callback(result)
        }
    }
}

fun JarvisClient.synthesizeSpeechAsync(text: String, voice: String = "en-US-AriaNeural", callback: (Result<VoiceResponse>) -> Unit) {
    GlobalScope.launch {
        val result = synthesizeSpeech(text, voice)
        withContext(Dispatchers.Main) {
            callback(result)
        }
    }
}

fun JarvisClient.authenticateAsync(deviceInfo: DeviceInfo, callback: (Result<AuthResponse>) -> Unit) {
    GlobalScope.launch {
        val result = authenticate(deviceInfo)
        withContext(Dispatchers.Main) {
            callback(result)
        }
    }
}

File: mobile-app-sdk/ios/JarvisSDK.swift
import Foundation
import AVFoundation
import Network

// MARK: - Models

public struct JarvisConfig {
    public let baseURL: String
    public let apiKey: String
    public let timeout: TimeInterval
    public let retryAttempts: Int
    public let enableLogging: Bool
    
    public init(
        baseURL: String,
        apiKey: String,
        timeout: TimeInterval = 30.0,
        retryAttempts: Int = 3,
        enableLogging: Bool = false
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.timeout = timeout
        self.retryAttempts = retryAttempts
        self.enableLogging = enableLogging
    }
}

public struct DeviceInfo {
    public let deviceID: String
    public let platform: Platform
    public let appVersion: String
    public let osVersion: String
    public let deviceModel: String
    public let timezone: String
    
    public init(
        deviceID: String,
        platform: Platform,
        appVersion: String,
        osVersion: String,
        deviceModel: String = UIDevice.current.model,
        timezone: String = TimeZone.current.identifier
    ) {
        self.deviceID = deviceID
        self.platform = platform
        self.appVersion = appVersion
        self.osVersion = osVersion
        self.deviceModel = deviceModel
        self.timezone = timezone
    }
}

public enum Platform: String, CaseIterable {
    case iOS = "ios"
    case android = "android"
}

public struct AuthenticationResult {
    public let success: Bool
    public let userID: String
    public let token: String
    public let sessionID: String
    public let expiresAt: Date
    public let userInfo: [String: Any]
}

public struct VoiceResponse {
    public let success: Bool
    public let text: String?
    public let audioData: Data?
    public let processingTime: TimeInterval
    public let confidence: Double
    public let language: String
}

public struct CommandResponse {
    public let success: Bool
    public let command: String
    public let response: String
    public let confidence: Double
    public let processingTime: TimeInterval
}

// MARK: - Errors

public enum JarvisError: Error, LocalizedError {
    case authenticationFailed(String)
    case networkError(String)
    case voiceProcessingError(String)
    case configurationError(String)
    case permissionDenied(String)
    case sessionExpired
    case invalidResponse
    
    public var errorDescription: String? {
        switch self {
        case .authenticationFailed(let message):
            return "Authentication failed: \(message)"
        case .networkError(let message):
            return "Network error: \(message)"
        case .voiceProcessingError(let message):
            return "Voice processing error: \(message)"
        case .configurationError(let message):
            return "Configuration error: \(message)"
        case .permissionDenied(let message):
            return "Permission denied: \(message)"
        case .sessionExpired:
            return "Session expired"
        case .invalidResponse:
            return "Invalid response from server"
        }
    }
}

// MARK: - Main SDK Class

public class JarvisClient {
    private var config: JarvisConfig?
    private var authToken: String?
    private var sessionID: String?
    private var userID: String?
    
    private let urlSession: URLSession
    private let audioEngine = AVAudioEngine()
    private let speechRecognizer = SFSpeechRecognizer()
    private let speechSynthesizer = AVSpeechSynthesizer()
    
    private var isRecording = false
    private var voiceSessionActive = false
    
    // Network monitoring
    private let networkMonitor = NWPathMonitor()
    private let networkQueue = DispatchQueue(label: "NetworkMonitor")
    
    public init() {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30.0
        configuration.timeoutIntervalForResource = 60.0
        self.urlSession = URLSession(configuration: configuration)
        
        setupNetworkMonitoring()
    }
    
    // MARK: - Configuration
    
    public func configure(config: JarvisConfig) {
        self.config = config
        if config.enableLogging {
            print("Jarvis SDK configured with base URL: \(config.baseURL)")
        }
    }
    
    // MARK: - Authentication
    
    public func authenticate(deviceInfo: DeviceInfo, completion: @escaping (Result<AuthenticationResult, JarvisError>) -> Void) {
        guard let config = config else {
            completion(.failure(.configurationError("SDK not configured")))
            return
        }
        
        let url = URL(string: "\(config.baseURL)/api/v2/mobile/auth/login")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(config.apiKey)", forHTTPHeaderField: "Authorization")
        
        let requestBody = [
            "device_info": [
                "device_id": deviceInfo.deviceID,
                "platform": deviceInfo.platform.rawValue,
                "app_version": deviceInfo.appVersion,
                "os_version": deviceInfo.osVersion,
                "device_model": deviceInfo.deviceModel,
                "timezone": deviceInfo.timezone
            ]
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(.configurationError("Failed to serialize request")))
            return
        }
        
        urlSession.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.handleAuthResponse(data: data, response: response, error: error, completion: completion)
            }
        }.resume()
    }
    
    private func handleAuthResponse(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<AuthenticationResult, JarvisError>) -> Void) {
        if let error = error {
            completion(.failure(.networkError(error.localizedDescription)))
            return
        }
        
        guard let httpResponse = response as? HTTPURLResponse else {
            completion(.failure(.invalidResponse))
            return
        }
        
        guard httpResponse.statusCode == 200 else {
            completion(.failure(.authenticationFailed("HTTP \(httpResponse.statusCode)")))
            return
        }
        
        guard let data = data else {
            completion(.failure(.invalidResponse))
            return
        }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            guard let success = json?["success"] as? Bool, success,
                  let userID = json?["user_id"] as? String,
                  let token = json?["mobile_token"] as? String,
                  let sessionID = json?["session_id"] as? String else {
                completion(.failure(.authenticationFailed("Invalid response format")))
                return
            }
            
            // Store authentication data
            self.authToken = token
            self.sessionID = sessionID
            self.userID = userID
            
            let expiresAt = Date().addingTimeInterval(30 * 24 * 60 * 60) // 30 days
            let userInfo = json?["user_info"] as? [String: Any] ?? [:]
            
            let result = AuthenticationResult(
                success: true,
                userID: userID,
                token: token,
                sessionID: sessionID,
                expiresAt: expiresAt,
                userInfo: userInfo
            )
            
            completion(.success(result))
            
        } catch {
            completion(.failure(.invalidResponse))
        }
    }
    
    // MARK: - Voice Processing
    
    public func startVoiceSession(completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        guard authToken != nil else {
            completion(.failure(.authenticationFailed("Not authenticated")))
            return
        }
        
        requestMicrophonePermission { [weak self] granted in
            if granted {
                self?.startRecording(completion: completion)
            } else {
                completion(.failure(.permissionDenied("Microphone access denied")))
            }
        }
    }
    
    private func requestMicrophonePermission(completion: @escaping (Bool) -> Void) {
        switch AVAudioSession.sharedInstance().recordPermission {
        case .granted:
            completion(true)
        case .denied:
            completion(false)
        case .undetermined:
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                DispatchQueue.main.async {
                    completion(granted)
                }
            }
        @unknown default:
            completion(false)
        }
    }
    
    private func startRecording(completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        do {
            let audioSession = AVAudioSession.sharedInstance()
            try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
            try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
            
            let inputNode = audioEngine.inputNode
            let recordingFormat = inputNode.outputFormat(forBus: 0)
            
            inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { [weak self] buffer, _ in
                self?.processAudioBuffer(buffer, completion: completion)
            }
            
            audioEngine.prepare()
            try audioEngine.start()
            
            isRecording = true
            voiceSessionActive = true
            
            if config?.enableLogging == true {
                print("Voice recording started")
            }
            
        } catch {
            completion(.failure(.voiceProcessingError("Failed to start recording: \(error.localizedDescription)")))
        }
    }
    
    private func processAudioBuffer(_ buffer: AVAudioPCMBuffer, completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        // Convert audio buffer to data
        guard let audioData = bufferToData(buffer) else {
            completion(.failure(.voiceProcessingError("Failed to convert audio buffer")))
            return
        }
        
        // Send to server for processing
        sendVoiceData(audioData, completion: completion)
    }
    
    private func bufferToData(_ buffer: AVAudioPCMBuffer) -> Data? {
        let audioBuffer = buffer.audioBufferList.pointee.mBuffers
        let data = Data(bytes: audioBuffer.mData!, count: Int(audioBuffer.mDataByteSize))
        return data
    }
    
    private func sendVoiceData(_ audioData: Data, completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        guard let config = config,
              let authToken = authToken,
              let sessionID = sessionID else {
            completion(.failure(.configurationError("SDK not properly configured")))
            return
        }
        
        let url = URL(string: "\(config.baseURL)/api/v2/mobile/voice/process")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        
        let base64Audio = audioData.base64EncodedString()
        let requestBody = [
            "session_id": sessionID,
            "audio_data": base64Audio,
            "language": "en",
            "format": "wav"
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(.voiceProcessingError("Failed to serialize request")))
            return
        }
        
        urlSession.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.handleVoiceResponse(data: data, response: response, error: error, completion: completion)
            }
        }.resume()
    }
    
    private func handleVoiceResponse(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        if let error = error {
            completion(.failure(.networkError(error.localizedDescription)))
            return
        }
        
        guard let httpResponse = response as? HTTPURLResponse else {
            completion(.failure(.invalidResponse))
            return
        }
        
        guard httpResponse.statusCode == 200 else {
            completion(.failure(.voiceProcessingError("HTTP \(httpResponse.statusCode)")))
            return
        }
        
        guard let data = data else {
            completion(.failure(.invalidResponse))
            return
        }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            guard let success = json?["success"] as? Bool, success else {
                completion(.failure(.voiceProcessingError("Voice processing failed")))
                return
            }
            
            let text = json?["response_text"] as? String
            let processingTime = json?["processing_time"] as? Double ?? 0.0
            let confidence = json?["confidence"] as? Double ?? 0.0
            let language = json?["language"] as? String ?? "en"
            
            let voiceResponse = VoiceResponse(
                success: true,
                text: text,
                audioData: nil,
                processingTime: processingTime,
                confidence: confidence,
                language: language
            )
            
            completion(.success(voiceResponse))
            
        } catch {
            completion(.failure(.invalidResponse))
        }
    }
    
    public func stopVoiceSession() {
        if isRecording {
            audioEngine.stop()
            audioEngine.inputNode.removeTap(onBus: 0)
            isRecording = false
            voiceSessionActive = false
            
            if config?.enableLogging == true {
                print("Voice recording stopped")
            }
        }
    }
    
    // MARK: - Command Processing
    
    public func executeCommand(_ command: String, completion: @escaping (Result<CommandResponse, JarvisError>) -> Void) {
        guard let config = config,
              let authToken = authToken,
              let sessionID = sessionID else {
            completion(.failure(.configurationError("SDK not properly configured")))
            return
        }
        
        let url = URL(string: "\(config.baseURL)/api/v2/mobile/command")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        
        let requestBody = [
            "session_id": sessionID,
            "command": command,
            "language": "en",
            "send_notification": false
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(.configurationError("Failed to serialize request")))
            return
        }
        
        urlSession.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.handleCommandResponse(data: data, response: response, error: error, completion: completion)
            }
        }.resume()
    }
    
    private func handleCommandResponse(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<CommandResponse, JarvisError>) -> Void) {
        if let error = error {
            completion(.failure(.networkError(error.localizedDescription)))
            return
        }
        
        guard let httpResponse = response as? HTTPURLResponse else {
            completion(.failure(.invalidResponse))
            return
        }
        
        guard httpResponse.statusCode == 200 else {
            completion(.failure(.networkError("HTTP \(httpResponse.statusCode)")))
            return
        }
        
        guard let data = data else {
            completion(.failure(.invalidResponse))
            return
        }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            guard let success = json?["success"] as? Bool, success else {
                completion(.failure(.voiceProcessingError("Command processing failed")))
                return
            }
            
            let command = json?["command"] as? String ?? ""
            let response = json?["response"] as? String ?? ""
            let confidence = json?["confidence"] as? Double ?? 0.0
            let processingTime = json?["processing_time"] as? Double ?? 0.0
            
            let commandResponse = CommandResponse(
                success: true,
                command: command,
                response: response,
                confidence: confidence,
                processingTime: processingTime
            )
            
            completion(.success(commandResponse))
            
        } catch {
            completion(.failure(.invalidResponse))
        }
    }
    
    // MARK: - Text-to-Speech
    
    public func synthesizeSpeech(text: String, voice: String = "en-US-AriaNeural", completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        guard let config = config,
              let authToken = authToken,
              let sessionID = sessionID else {
            completion(.failure(.configurationError("SDK not properly configured")))
            return
        }
        
        let url = URL(string: "\(config.baseURL)/api/v2/mobile/voice/process")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        
        let requestBody = [
            "session_id": sessionID,
            "text": text,
            "voice_settings": [
                "voice": voice
            ],
            "language": "en"
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(.configurationError("Failed to serialize request")))
            return
        }
        
        urlSession.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.handleTTSResponse(data: data, response: response, error: error, completion: completion)
            }
        }.resume()
    }
    
    private func handleTTSResponse(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<VoiceResponse, JarvisError>) -> Void) {
        if let error = error {
            completion(.failure(.networkError(error.localizedDescription)))
            return
        }
        
        guard let httpResponse = response as? HTTPURLResponse else {
            completion(.failure(.invalidResponse))
            return
        }
        
        guard httpResponse.statusCode == 200 else {
            completion(.failure(.voiceProcessingError("HTTP \(httpResponse.statusCode)")))
            return
        }
        
        guard let data = data else {
            completion(.failure(.invalidResponse))
            return
        }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            guard let success = json?["success"] as? Bool, success else {
                completion(.failure(.voiceProcessingError("TTS processing failed")))
                return
            }
            
            let text = json?["response_text"] as? String
            let processingTime = json?["processing_time"] as? Double ?? 0.0
            let confidence = json?["confidence"] as? Double ?? 1.0
            let language = json?["language"] as? String ?? "en"
            
            var audioData: Data?
            if let base64Audio = json?["response_audio"] as? String {
                audioData = Data(base64Encoded: base64Audio)
            }
            
            let voiceResponse = VoiceResponse(
                success: true,
                text: text,
                audioData: audioData,
                processingTime: processingTime,
                confidence: confidence,
                language: language
            )
            
            completion(.success(voiceResponse))
            
        } catch {
            completion(.failure(.invalidResponse))
        }
    }
    
    // MARK: - Network Monitoring
    
    private func setupNetworkMonitoring() {
        networkMonitor.pathUpdateHandler = { [weak self] path in
            if path.status == .satisfied {
                if self?.config?.enableLogging == true {
                    print("Network connection available")
                }
            } else {
                if self?.config?.enableLogging == true {
                    print("Network connection lost")
                }
            }
        }
        
        networkMonitor.start(queue: networkQueue)
    }
    
    // MARK: - Cleanup
    
    deinit {
        stopVoiceSession()
        networkMonitor.cancel()
    }
}

// MARK: - Extensions

extension JarvisClient {
    public func isAuthenticated() -> Bool {
        return authToken != nil && userID != nil
    }
    
    public func getCurrentUserID() -> String? {
        return userID
    }
    
    public func getCurrentSessionID() -> String? {
        return sessionID
    }
    
    public func logout() {
        authToken = nil
        sessionID = nil
        userID = nil
        stopVoiceSession()
    }
}